### PR TITLE
Rebinding Chat Hotkeys

### LINKED ISSUE #17446: Auto-completion issues
When using auto-completion it can be a bit annoying when you want to toggle team/all chat - you have to erase everything (I believe pressing `Home` would also work though I haven't tested it). Perhaps we should change either toggle or autocomplete (Ctrl+Tab/Space?) to another key?

### PR DESCRIPTION
**Pull request by abmyii**
Fixes #17446.

### CODE REVIEW

**Comment by dragunoff** [comment]
The basic approach here is OK as already discussed in #17446. However since this is changing a well established hotkey (<kbd>Tab</kbd> for toggling chat mode) and is adding a new hotkey (<kbd>Shift</kbd> + <kbd>Enter</kbd> for opening "all" chat) I'd like to do something about improving the discoverability of these keys :thinking: 

For the hotkeys that open the chat I suggest making them into real hotkeys (not hard-coded). Here's the gist of how to do that:
https://github.com/dragunoff/OpenRA/commit/02f24eae764a970e0d93a2ad0e138947ed2f4929

For the mode toggle I suggest to simply add a tooltip to the "Team/All" button that simply shows the hard-coded key combination. A new tooltip template has to be created (using `LabelWithHighlightWidget` for the TEXT part of the tooltip) so that it looks like this:
image

For the auto-complete hotkey just leave it as-is because <kbd>Tab</kbd> is a widely used key for that across software.

**Comment by abmyii** [comment]
> For the hotkeys that open the chat I suggest making them into real hotkeys (not hard-coded). Here's the gist of how to do that

Thanks, I'll see what I can do with that!

> For the mode toggle I suggest to simply add a tooltip to the "Team/All" button that simply shows the hard-coded key combination. A new tooltip template has to be created (using LabelWithHighlightWidget for the TEXT part of the tooltip) so that it looks like this:

I agree - that is the most straightforward solution. I'll give that a shot and see how it goes.

**Comment by abmyii** [comment]
@dragunoff I added the following to `BUTTON@CHAT_MODE` (https://github.com/OpenRA/OpenRA/blob/bleed/mods/common/chrome/ingame-chat.yaml#L22), yet I don't see any tooltips. Any idea why?
```yaml
TooltipContainer: TOOLTIP_CONTAINER
TooltipText: Enter/Shift+Enter for team/all chat
```

**Comment by dragunoff** [comment]
> @dragunoff I added the following to `BUTTON@CHAT_MODE` (https://github.com/OpenRA/OpenRA/blob/bleed/mods/common/chrome/ingame-chat.yaml#L22), yet I don't see any tooltips. Any idea why?
> 
> ```yaml
> TooltipContainer: TOOLTIP_CONTAINER
> TooltipText: Enter/Shift+Enter for team/all chat
> ```

That should work :thinking: Perhaps you were testing in TD? Keep in mind that TD has it's own set of chrome yaml files for most things (so there is a file `mods/cnc/chrome/ingame-chat.yaml` for example). 

However I think we need to fake the real hotkey tooltip by using a new template that has a `LabelWithHighlight` for the tooltip text. Here's how to do that: https://github.com/dragunoff/OpenRA/commit/f19d877a6797e57cc1f94db035baca7816454e08. Also the tooltip should highlight the `Shift + Tab` combo as that is what this button does. The `Enter` and `Shift + Enter` hotkeys are already displayed in the hotkey settings.

**Comment by abmyii** [comment]
> That should work thinking Perhaps you were testing in TD? Keep in mind that TD has it's own set of chrome yaml files for most things (so there is a file `mods/cnc/chrome/ingame-chat.yaml` for example).

No, I was using CNC. I didn't modify the `tooltips.yaml` like you did (https://github.com/dragunoff/OpenRA/commit/f19d877a6797e57cc1f94db035baca7816454e08#diff-8e162ea17ab3fd3a11013963ca33fc1edc5baf52ec2a7b973a4afd6fdd026a94) though.

**Comment by abmyii** [comment]
> Here's how to do that: [dragunoff@f19d877](https://github.com/dragunoff/OpenRA/commit/f19d877a6797e57cc1f94db035baca7816454e08).

Haha - you've just done all of the work! What else is there left to do?

**Comment by dragunoff** [comment]
> 
> Haha - you've just done all of the work! What else is there left to do?

It has to be added to the other mods (I only covered TD). So feel free to pick my work to your branch and finish it. I wanted to show you how some things work with the UI in OpenRA as I know it can be daunting looking at the pile of chrome YAML.

**Comment by abmyii** [comment]
> > Haha - you've just done all of the work! What else is there left to do?
> 
> It has to be added to the other mods (I only covered TD). So feel free to pick my work to your branch and finish it. I wanted to show you how some things work with the UI in OpenRA as I know it can be daunting looking at the pile of chrome YAML.

OK, will do. Appreciate the introduction - I do find it quite overwhelming!

**Comment by abmyii** [comment]
I'm sure there's a good reason for it, but `mods/cnc/chrome/ingame-chat.yaml` is identical to `mods/common/chrome/ingame-chat.yaml` (ignoring the changes that you added). Why doesn't it just inherit the `common` one, instead of being a duplicate, and should I delete it and use the common one instead?

**Comment by dragunoff** [comment]
> I'm sure there's a good reason for it, but `mods/cnc/chrome/ingame-chat.yaml` is identical to `mods/common/chrome/ingame-chat.yaml` (ignoring the changes that you added). Why doesn't it just inherit the `common` one, instead of being a duplicate, and should I delete it and use the common one instead?

I guess it was just copied from the common one at one point because TD has a distinct UI. There may have been differences in the file along the way. I won't dig into the history now but for this PR I'd say leave it as it is.

File: mods/common/chrome/settings-hotkeys.yaml
Comment (dragunoff) [inline mods/common/chrome/settings-hotkeys.yaml:30]: Please add this to `mods/cnc/chrome/settings-hotkeys.yaml` as well.

File: mods/common/chrome/tooltips.yaml
Comment (dragunoff) [inline mods/common/chrome/tooltips.yaml]: This should match the height of the other button tooltips, 29 in this case.

File: mods/d2k/chrome/tooltips.yaml
Comment (dragunoff) [inline mods/d2k/chrome/tooltips.yaml]: This should match the height of the other button tooltips, 31 in this case.

File: mods/d2k/chrome/tooltips.yaml
Comment (dragunoff) [inline mods/d2k/chrome/tooltips.yaml]: This should match the other button tooltips, `X: 5` and `Y: 4`.

File: mods/common/chrome/tooltips.yaml
Comment (dragunoff) [inline mods/common/chrome/tooltips.yaml]: This should match the other button tooltips, `X: 7` and `Y: 3`.

File: mods/cnc/chrome/ingame-chat.yaml
Comment (dragunoff) [inline mods/cnc/chrome/ingame-chat.yaml]: Please revert this change to preserve the empty line at the end of the file.

File: mods/common/hotkeys/game.yaml
Comment (dragunoff) [inline mods/common/hotkeys/game.yaml]: Please revert this change to preserve the empty line at the end of the file.

File: mods/cnc/chrome/ingame-chat.yaml
Comment (dragunoff) [inline mods/cnc/chrome/ingame-chat.yaml]: Please also add this to the button in the lobby chat.

File: mods/common/chrome/ingame-chat.yaml
Comment (dragunoff) [inline mods/common/chrome/ingame-chat.yaml]: Please also add this to the button in the lobby chat.

**Review by dragunoff — CHANGES_REQUESTED** [review]
I have noted a few things that need to be changed. See my comments on the code. 

Additionally there is one important thing to address - make the "Enter" key a valid hotkey entry (because now it's not possible for the user to bind to "Enter" from the UI). For this you basically need to change line 90 in `HotkeyEntryWidget.cs` (remove `Keycode.RETURN, Keycode.KP_ENTER` from the list of ignored keys).

Another minor thing to address is the dangling tooltip that remains after closing the chat. To reproduce open the in-game chat, point to the toggle chat button to trigger the tooltip and the close the chat by pressing "Enter". You should continue to see the tooltip even when the button is gone. To fix this simply add `Ui.ResetTooltips();` to the end of `IngameChatLogic.CloseChat()`.

I noticed a small visual problem with the new tooltip template - the background is a little wider than the text:
image

I think the above is happening because the curly braces that denote the text to highlight are also measured as part of the text. This is not a blocker for this PR and can be fixed separately.

**Comment by dragunoff** [comment]
@abmyii Looks good but please also take care of the other two points that I raised in https://github.com/OpenRA/OpenRA/pull/19460#pullrequestreview-692827166 (about `HotkeyEntryWidget` and `Ui.ResetTooltips`).

**Comment by abmyii** [comment]
> @abmyii Looks good but please also take care of the other two points that I raised in [#19460 (review)](https://github.com/OpenRA/OpenRA/pull/19460#pullrequestreview-692827166) (about `HotkeyEntryWidget` and `Ui.ResetTooltips`).

Decided to work on a fix for the label being too large before pushing those changes which I had already made yesterday. Regarding the fix, I don't think it's the most proper way, but I couldn't really think of a better method due to my lack of familiarity with C#. I guess I could do something like the following, but it seems a bit messy:

```C#
var labelWidth = font.Measure(label is LabelWithHighlightWidget ? text.Replace("{", "").Replace("}", "") : text).X;
```

**Comment by dragunoff** [comment]
@abmyii Using a ternary is fine, you can split it into multiple lines to make it more readable, something like:
```csharp
var labelWidth = label is LabelWithHighlightWidget ?
	font.Measure(text.Replace("{", "").Replace("}", "")).X :
	font.Measure(text).X;
```

However this solves the measurement issue only for button tooltips. And the tooltip logic should not have to deal with that. I think this bug calls for a more comprehensive solution. Perhaps by introducing a method that does the text measurement in the label widget itself. I think we should leave this to a follow up PR instead of scope creeping this one. I suggest we open an issue to track the bug and then drop the last commit from here.

**Comment by abmyii** [comment]
> @abmyii Using a ternary is fine, you can split it into multiple lines to make it more readable, something like:
> 
> ```cs
> var labelWidth = label is LabelWithHighlightWidget ?
> 	font.Measure(text.Replace("{", "").Replace("}", "")).X :
> 	font.Measure(text).X;
> ```
> 
> However this solves the measurement issue only for button tooltips. And the tooltip logic should not have to deal with that. I think this bug calls for a more comprehensive solution. Perhaps by introducing a method that does the text measurement in the label widget itself. I think we should leave this to a follow up PR instead of scope creeping this one. I suggest we open an issue to track the bug and then drop the last commit from here.

Done. Can you make the issue or shall I?

**Comment by pchote** [comment]
I remember solving the highlighted text measurements a few weeks ago for another PR/branch. Are you sure we haven't already merged it?

**Comment by abmyii** [comment]
> I remember solving the highlighted text measurements a few weeks ago for another PR/branch. Are you sure we haven't already merged it?

No idea - how can I check?

**Comment by pchote** [comment]
Ah, it was in #19324 which I decided to not continue with.

I'll split 5ba16284d992bea4d96d170c41e0070ff5e8d5e1 out into its own PR so you can use it here. As you can see, doing this properly isn't as simple as just removing <kbd>{</kbd> and <kbd>}</kbd>, as there are various cases where these characters will be rendered in highlighted strings.

**Comment by abmyii** [comment]
> Ah, it was in #19324 which I decided to not continue with.

That's a shame - it looked very promising.

> I'll split [5ba1628](https://github.com/OpenRA/OpenRA/commit/5ba16284d992bea4d96d170c41e0070ff5e8d5e1) out into its own PR so you can use it here. As you can see, doing this properly isn't as simple as just removing { and }, as there are various cases where these characters will be rendered in highlighted strings.

Will I have to pull it into my branch? If so, how?

**Comment by dragunoff** [comment]
> 
> Will I have to pull it into my branch? If so, how?

@abmyii pchote will open a new PR with the fix and then you can rebase your branch onto his. But you don't need to worry too much about that because the bug is purely visual. Until then you can prepare this branch for merging by squashing the commits.

**Comment by abmyii** [comment]
> > Will I have to pull it into my branch? If so, how?
> 
> @abmyii pchote will open a new PR with the fix and then you can rebase your branch onto his. But you don't need to worry too much about that because the bug is purely visual. Until then you can prepare this branch for merging by squashing the commits.

OK, done.

**Comment by dragunoff** [comment]
@abmyii Please also tidy up the commit message - this is the first thing we see in the project history and it should be meaningful.

**Comment by abmyii** [comment]
> @abmyii Please also tidy up the commit message - this is the first thing we see in the project history and it should be meaningful.

Is that alright? I'm not great with commit message convention, sorry.

**Comment by dragunoff** [comment]
Yes, this is better though I think you can make the heading shorter and use the body to provide additional info.

The commit message has a heading and a body part. The heading is the first line and is what we see most of the time. Ideally it should be kept shorter (for scanability and to avoid truncation by github and other software). The body part is what comes in the next lines and is typically used to elaborate on more complex changes.

**Comment by pchote** [comment]
Looking at this more closely, it would be better to declare the `CHAT_MODE` tooltip using
```
Key: Tab SHIFT
TooltipText: Toggle chat mode
TooltipContainer: TOOLTIP_CONTAINER
TooltipTemplate: BUTTON_TOOLTIP
```

We can then avoid the new `BUTTON_WITH_LABEL_HIGHLIGHT_TOOLTIP` definitions and don't need to worry about measuring highlighted strings at all.

**Comment by dragunoff** [comment]
> Looking at this more closely, it would be better to declare the `CHAT_MODE` tooltip using [...]
> 

This is what I did at first but it felt like a hack. I'm not strongly against it though.

**Comment by pchote** [comment]
It could be made cleaner by changing `TextFieldWidget.OnTabKey` to be called with a `KeyInput` argument. This would be best done in its own commit, changing also `OnEnterKey` and `OnEscKey` for consistency even though those aren't directly relevant to this PR.


`LobbyLogic` can then do
```csharp
chatTextField.OnTabKey = e =>
{
	if (!chatMode.Key.IsActivatedBy(e) || chatMode.IsDisabled())
	{
		chatTextField.Text = tabCompletion.Complete(chatTextField.Text);
		chatTextField.CursorPosition = chatTextField.Text.Length;
	}
	else
		chatMode.OnKeyPress(e);

	return true;
};
```
(and the now redundant `SwitchTeamChat` method can be removed), and similarly for `IngameChatLogic`.

File: OpenRA.Mods.Common/Widgets/TextFieldWidget.cs
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/TextFieldWidget.cs]: Pass `_` here as well for consistency with the other methods.

File: mods/cnc/chrome/ingame-chat.yaml
Comment (dragunoff) [inline mods/cnc/chrome/ingame-chat.yaml]: `BUTTON_TOOLTIP` is the default value so this is redundant.

File: mods/cnc/chrome/lobby.yaml
Comment (dragunoff) [inline mods/cnc/chrome/lobby.yaml]: `BUTTON_TOOLTIP` - same here.

File: mods/common/chrome/ingame-chat.yaml
Comment (dragunoff) [inline mods/common/chrome/ingame-chat.yaml]: `BUTTON_TOOLTIP` - same here.

File: mods/common/chrome/lobby.yaml
Comment (dragunoff) [inline mods/common/chrome/lobby.yaml]: `BUTTON_TOOLTIP` - same here.

**Review by dragunoff — CHANGES_REQUESTED** [review]

**Review by dragunoff — DISMISSED** [review]

File: mods/common/hotkeys/chat.yaml
Comment (pchote) [inline mods/common/hotkeys/chat.yaml]: TBH I would struggle to guess what "All Chat" meant if it weren't for the context of this PR. Can we please come up with new clearer names for these hotkeys?

**Review by pchote — COMMENTED** [review]
The code changes look reasonable to me now. I have one comment about UX, and the commits still need to be reworked so the `KeyInput` changes are their own distinct first commit.

File: mods/common/hotkeys/chat.yaml
Comment (abcdefg30) [reply mods/common/hotkeys/chat.yaml]: If "All" isn't clear we could name it "General Chat"?

Comment (pchote) [reply mods/common/hotkeys/chat.yaml]: If this is the general chat, shouldn't it just be "Chat" and the other one have the qualifier?

Comment (dragunoff) [reply mods/common/hotkeys/chat.yaml]: > If this is the general chat, shouldn't it just be "Chat" and the other one have the qualifier?

Yes, "Open Chat" and "Open Team Chat" seem like a far better choice. I didn't think this through when I suggested the names to @abmyii :+1:

Comment (dragunoff) [reply mods/common/hotkeys/chat.yaml]: I remember why I proposed this naming - the "Open Team Chat" key actually opens the general chat if the player is not on a team. It serves both functions and that's why I gave the qualifier to the other one :thinking: So maybe "Open Chat" and "Open General Chat" is better after all?

Comment (abcdefg30) [reply mods/common/hotkeys/chat.yaml]: Imo "Open Team Chat" opening the general chat if you're not on a team is ok behavioral-wise. (What else would a player expect the key to do when not on a team?)

Comment (abmyii) [reply mods/common/hotkeys/chat.yaml]: How about `Open Primary Chat` and `Open (General) Chat`?

Comment (pchote) [reply mods/common/hotkeys/chat.yaml]: > Imo "Open Team Chat" opening the general chat if you're not on a team is ok behavioral-wise. (What else would a player expect the key to do when not on a team?)

I agree. How do other games that have separate bindings for opening team/all chat handle this?

Comment (dragunoff) [reply mods/common/hotkeys/chat.yaml]: > 
> I agree. How do other games that have separate bindings for opening team/all chat handle this?

I did some research and the most wide-spread configuration seems to be <kbd>Enter</kbd> for team chat, <kbd>Shift</kbd> + <kbd>Enter</kbd> for general chat and <kbd>Tab</kbd> to toggle. Second and third gen CNC games have it as <kbd>Backspace</kbd> for team and <kbd>Enter</kbd> for all. There are some funkier setups like [the prefix system in AoE II](https://gaming.stackexchange.com/a/343641). However I've not been able to see how the "team chat" key behaves when not on a team because it seems I need a real multi-player game to test that. 

After all I think the behavior we've set up here is good and we just need to deal with the labeling problem. How about we make it all explicit: "Open Team Chat" and "Open General Chat"?

Comment (pchote) [reply mods/common/hotkeys/chat.yaml]: > How about we make it all explicit: "Open Team Chat" and "Open General Chat"?

This sounds good to me 👍

Comment (abmyii) [reply mods/common/hotkeys/chat.yaml]: > After all I think the behavior we've set up here is good and we just need to deal with the labeling problem. How about we make it all explicit: "Open Team Chat" and "Open General Chat"?

So that means that the Open Team Chat hotkey (Enter) will only work in team games and will not default to `All` chat?

Comment (dragunoff) [reply mods/common/hotkeys/chat.yaml]: > 
> So that means that the Open Team Chat hotkey (Enter) will only work in team games and will not default to `All` chat?

No, functionality remains the same ("team" defaults to "all" if not on a team) - we're only changing the labels and hotkey names.

**Review by pchote — APPROVED** [review]
LGTM now. It would be good if @dragunoff can reconfirm their :+1: before merging.

**Review by dragunoff — APPROVED** [review]
LGTM :+1: 

I think it would be good to add the hard-coded "Autocomplete" and "Toggle chat mode" keys to the hotkeys list as readonly entries. This is of course work for a follow up PR.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/867f31fecd2b457088485d5e49d2e202f4fd8252

**Pull request merged**
